### PR TITLE
fix(api): honour pageNumber in /api/studies pagination

### PR DESCRIPTION
**Pull request by Pradyuman-aviator**
The `getAllStudies` endpoint accepted `pageSize` and `pageNumber` parameters but silently ignored `pageNumber`, always returning the first page via `stream().limit(pageSize)`.

Replace with a proper offset-based `subList`:
- `pageNumber` is 1-based; `null` and `0` are treated as page 1 for backward compatibility with `PagingConstants.MIN_PAGE_NUMBER=0`
- Long arithmetic prevents int overflow for large `pageSize * pageNumber`
- Documents the in-memory pagination trade-off: `@PostFilter` (which enforces per-study authorization) runs after the DB fetch, so the authorized study set is unknown at query time —> DB-level `LIMIT/OFFSET` is deferred to a follow-up once the auth layer is restructured

Describe changes proposed in this pull request:
- Fix `pageNumber` being silently ignored in `ColumnStoreStudyController.getAllStudies` — callers requesting page 2+ always received page 1
- Use `subList(fromIndex, toIndex)` with overflow-safe long arithmetic instead of `stream().limit()`
- Add `ColumnStoreStudyControllerTest` with 7 cases: no pagination, `pageNumber=0` backward compat, explicit page 1, page 2 (primary regression), partial last page, and page beyond end

# Checks
- [x] The commit log is comprehensible. It follows [7 rules of great commit messages](http://chris.beams.io/posts/git-commit/). We can fix this during merge by using a squash+merge if necessary
- [x] Has tests — `ColumnStoreStudyControllerTest` with 7 unit test cases
- [x] Is this PR adding logic based on one or more **clinical** attributes? No.
- [x] Label: Bug Fix

### CODE REVIEW

**Comment by bitdive-review** [comment]
## BitDive Runtime Review - PR #12274

**PR**: https://github.com/cBioPortal/cbioportal/pull/12274  
**Branch**: `perf/fix-study-pagination-page-number-ignored` -> `master`  
**Method**: BitDive runtime trace comparison + direct HTTP verification + targeted code inspection  
**Verdict**: **APPROVE WITH NOTES**  
**Confidence**: High — bug reproduction and fix demonstrated across 3 runtime scenarios with 4 additional stable-flow confirmations

### Summary

This PR fixes a single-method pagination bug in `ColumnStoreStudyController.getAllStudies` where `pageNumber` was silently ignored — `stream().limit(pageSize)` always returned the first page regardless of which page was requested. The fix replaces this with `subList(fromIndex, toIndex)` using 1-based offset arithmetic (`pageSize * (pageNumber - 1)`), `Math.min` clamping against list size, and `long` intermediate arithmetic to prevent integer overflow. Three independent runtime scenarios confirm the bug on base and the fix on head: page 2 now returns the correct empty result (vs base returning first-page data), page beyond end returns empty without `IndexOutOfBoundsException` (vs base returning data), and the META projection's `X-Total-Count` header now reports the post-pagination count (0 vs base's 1). Four backward-compatibility paths (explicit page 1, null `pageNumber`, `pageNumber=0`, no pagination) show zero contract divergence. Seven unit tests with 25 mocked studies cover multi-study offset arithmetic unreachable by the single-study runtime seed. The change is a read-only GET endpoint — no writes, no downstream service calls, minimal blast radius.

### Runtime Verification Matrix

| Area | Result | Behavior Δ | Meaning | Evidence |
| --- | --- | --- | --- | --- |
| Page 2 request (SUMMARY) | Fixed | `HTTP 200 both sides · body length 1→0 · zero writes · zero downstream` | pageNumber now honored — page 2 no longer returns page-1 data | [before](https://cloud.bitdive.io/monitoring-api/public/trace/share?callId=43d62ca4-825b-11f1-a26a-9def06531e5b) / [after](https://cloud.bitdive.io/monitoring-api/public/trace/share?callId=d3480372-825b-11f1-91b8-65dd1cfa6f22) |
| Page beyond end (pageSize=100) | Fixed | `HTTP 200 both sides · body length 1→0 · Math.min clamping prevents IOOBE · zero writes` | Beyond-end pages return empty, not first-page overflow | [before](https://cloud.bitdive.io/monitoring-api/public/trace/share?callId=44225101-825b-11f1-a26a-9def06531e5b) / [after](https://cloud.bitdive.io/monitoring-api/public/trace/share?callId=d39e12df-825b-11f1-91b8-65dd1cfa6f22) |
| META projection (page 2 header) | Fixed | `HTTP 200 both sides · X-Total-Count 1→0 · total-count 1→0 · body empty both sides` | META count header reflects post-pagination slice | [before](https://cloud.bitdive.io/monitoring-api/public/trace/share?callId=4432569a-825b-11f1-a26a-9def06531e5b) / [after](https://cloud.bitdive.io/monitoring-api/public/trace/share?callId=d3ae1878-825b-11f1-91b8-65dd1cfa6f22) |
| Explicit page 1 | Stable | `zero contract drift · body 1=1 · study_es_0 both sides` | Page 1 path unchanged | [before](https://cloud.bitdive.io/monitoring-api/public/trace/share?callId=43e547dd-825b-11f1-a26a-9def06531e5b) / [after](https://cloud.bitdive.io/monitoring-api/public/trace/share?callId=d35b3d5b-825b-11f1-91b8-65dd1cfa6f22) |
| Null pageNumber default | Stable | `zero contract drift · defaults to page 1 · body 1=1` | Null handling unchanged | [before](https://cloud.bitdive.io/monitoring-api/public/trace/share?callId=43f46316-825b-11f1-a26a-9def06531e5b) / [after](https://cloud.bitdive.io/monitoring-api/public/trace/share?callId=d36e7744-825b-11f1-91b8-65dd1cfa6f22) |
| pageNumber=0 backward compat | Stable | `zero contract drift · treated as page 1 · body 1=1` | @Min(0) boundary unchanged | [before](https://cloud.bitdive.io/monitoring-api/public/trace/share?callId=44037e4f-825b-11f1-a26a-9def06531e5b) / [after](https://cloud.bitdive.io/monitoring-api/public/trace/share?callId=d37e55cd-825b-11f1-91b8-65dd1cfa6f22) |
| No pagination path | Stable | `zero contract drift · pageSize=null branch untouched · body 1=1` | pageSize=null returns full list | [before](https://cloud.bitdive.io/monitoring-api/public/trace/share?callId=44129988-825b-11f1-a26a-9def06531e5b) / [after](https://cloud.bitdive.io/monitoring-api/public/trace/share?callId=d38e5b66-825b-11f1-91b8-65dd1cfa6f22) |
| Unit tests (7 cases, 25 mocked studies) | Verified | `Tests 7/7 pass · page 2 returns study-10..19 · partial page returns study-20..24 · beyond-end returns empty` | Multi-study offset covered | Tests in PR |

<details>
<summary>How to read BitDive evidence</summary>

**Behavior Δ** (matrix) and the red/green `diff` Behavior contract inside each Change summarize the runtime contract.  
**Evidence** links open the full BitDive share: call tree · headers · body · downstream · first divergence.

- **Trace pair** — full before/after JVM trees compared
- **Code only** — unit test inspection (no JVM trace for Maven runs)
- **Replay** — no replay regression data for this PR

</details>

<details>
<summary>Review scope and validation coverage</summary>

```text
cbioportal
  -> ColumnStoreStudyController.getAllStudies()           [single-method fix]
     -> GetCancerStudyMetadataUseCase.execute()
        -> ClickhouseCancerStudyRepository.getCancerStudiesMetadataSummary()

Test coverage (new in PR):
  -> ColumnStoreStudyControllerTest                        [7 cases, 25 mocked studies]
```

| Behavior | Scenario | Evidence | Status |
| --- | --- | --- | --- |
| pageNumber honored (SUMMARY) | page 2, pageSize=1 | Trace pair | **Fix proven** |
| Math.min clamping | page beyond end, pageSize=100 | Trace pair | **Fix proven** |
| META projection X-Total-Count | page 2, META | Trace pair | **Fix proven** |
| Explicit page 1 | pageNumber=1 | Trace pair | Stable |
| Null pageNumber default | pageNumber omitted | Trace pair | Stable |
| pageNumber=0 boundary | @Min(0) allowed value | Trace pair | Stable |
| No-pagination path | pageSize=null | Trace pair | Stable |
| Multi-study offset correctness | 25 mocked studies | Unit tests | Code-verified |

**Out of scope:** ClickHouse-level pagination — the repository also ignores `pageNumber` (pre-existing, explicitly deferred per PR description due to the `@PostFilter` authorization constraint).

</details>

<details>
<summary>Change #1 — pageNumber honored in /api/studies pagination</summary>

**File:** `src/main/java/org/cbioportal/application/rest/vcolumnstore/ColumnStoreStudyController.java` (line ~125)

The controller previously ignored `pageNumber` entirely: `studies.stream().limit(pageSize).toList()` always returned the first page. The fix uses `subList(fromIndex, toIndex)` with a 1-based offset (`pageSize * (pageNumber - 1)`), `Math.min` clamping against `studies.size()` to prevent `IndexOutOfBoundsException`, and `long` intermediate arithmetic to avoid integer overflow on large products.

#### Behavior contract

```diff
# GET /api/studies?projection=SUMMARY&pageSize=1&pageNumber=2 — runtime contract
- HTTP 200 · body=[study_es_0] (1 item, bug: pageNumber ignored)
+ HTTP 200 · body=[] (0 items, fix: offset = 1*(2-1) = 1, subList(1,1) = empty)
- studies.stream().limit(pageSize).toList()
+ studies.subList(fromIndex, toIndex) with Math.min clamping
- (no SQL writes — read-only CACHE+METHOD)
+ (no SQL writes — read-only CACHE+METHOD)
- (no downstream REST)
+ (no downstream REST)
- data: page 2 returns first-page data
+ data: page 2 returns the correct empty slice (single-study seed)
# first divergence: ColumnStoreStudyController.getAllStudies() return
- useCase.execute() → 1 study → stream().limit(1) → ResponseEntity(study_es_0)
+ useCase.execute() → 1 study → subList(1, 1) → ResponseEntity() (empty)
```

#### Scenario matrix

| Scenario | Request | Before PR | After PR | Purpose |
| --- | --- | --- | --- | --- |
| Page 2 (primary regression) | `pageSize=1&pageNumber=2` | `200`, `[study_es_0]` | `200`, `[]` | Prove pageNumber now honored |
| Page beyond end | `pageSize=100&pageNumber=2` | `200`, `[study_es_0]` | `200`, `[]` | Math.min clamping, no IOOBE |
| META X-Total-Count | `projection=META&pageSize=1&pageNumber=2` | `200`, `X-Total-Count: 1` | `200`, `X-Total-Count: 0` | Header reflects post-pagination count |

#### Trace evidence

| Trace | Scenario | HTTP | Body / Header | Downstream | Result |
| --- | --- | --- | --- | --- | --- |
| [before page2](https://cloud.bitdive.io/monitoring-api/public/trace/share?callId=43d62ca4-825b-11f1-a26a-9def06531e5b) | page 2 SUMMARY | `200` | `[study_es_0]` | none | Bug reproduced |
| [after page2](https://cloud.bitdive.io/monitoring-api/public/trace/share?callId=d3480372-825b-11f1-91b8-65dd1cfa6f22) | page 2 SUMMARY | `200` | `[]` | none | Fix demonstrated |
| [before beyond-end](https://cloud.bitdive.io/monitoring-api/public/trace/share?callId=44225101-825b-11f1-a26a-9def06531e5b) | pageSize=100 page 2 | `200` | `[study_es_0]` | none | Bug reproduced |
| [after beyond-end](https://cloud.bitdive.io/monitoring-api/public/trace/share?callId=d39e12df-825b-11f1-91b8-65dd1cfa6f22) | pageSize=100 page 2 | `200` | `[]` | none | Clamping proven |
| [before META](https://cloud.bitdive.io/monitoring-api/public/trace/share?callId=4432569a-825b-11f1-a26a-9def06531e5b) | META page 2 | `200` | `X-Total-Count: 1` | none | Bug reproduced |
| [after META](https://cloud.bitdive.io/monitoring-api/public/trace/share?callId=d3ae1878-825b-11f1-91b8-65dd1cfa6f22) | META page 2 | `200` | `X-Total-Count: 0` | none | Fix demonstrated |

#### Contract delta

| Layer | Before PR | After PR | Result |
| --- | --- | --- | --- |
| HTTP | `200` | `200` | Unchanged (correct behavior, not an error path) |
| Body (SUMMARY) | `[study_es_0]` on page 2 (bug) | `[]` on page 2 (correct) | Fixed |
| Header (META) | `X-Total-Count: 1` on page 2 (bug) | `X-Total-Count: 0` on page 2 (correct) | Fixed |
| Persistence | read-only CACHE+METHOD | read-only CACHE+METHOD | Unchanged |
| Downstream | none | none | Unchanged |
| Overflow safety | n/a (limit only) | `long` arithmetic + `Math.min` clamping | Hardened |

#### Key trace delta

First meaningful divergence: `ColumnStoreStudyController.getAllStudies()` return.

| Branch | Trace shape |
| --- | --- |
| Before PR | `useCase.execute()` → 1 study → `stream().limit(1)` → `ResponseEntity(study_es_0)` |
| After PR | `useCase.execute()` → 1 study → `subList(1, 1)` → `ResponseEntity()` (empty) |

The use case and repository behave identically on both sides — the sole divergence is the controller-level pagination slice.

</details>

<details>
<summary>Change #2 — Backward compatibility paths stable</summary>

All four stable-flow scenarios show **CLEAN baselines** with zero contract divergence (no first divergence, no write delta, no downstream delta):

| Flow | Before / After | Divergence |
| --- | --- | --- |
| Explicit page 1 (`pageNumber=1`) | `200`, `[study_es_0]` / `200`, `[study_es_0]` | none |
| Null `pageNumber` default | `200`, `[study_es_0]` / `200`, `[study_es_0]` | none |
| `pageNumber=0` boundary | `200`, `[study_es_0]` / `200`, `[study_es_0]` | none |
| No pagination (`pageSize=null`) | `200`, `[study_es_0]` / `200`, `[study_es_0]` | none |

The `pageNumber <= 0 || pageNumber == null` short-circuit preserves the prior first-page behavior, and the `pageSize == null` branch is untouched.

</details>

<details>
<summary>Unit test coverage — multi-study offset arithmetic</summary>

`ColumnStoreStudyControllerTest` — 7/7 pass, 25 mocked studies covering edge cases unreachable by the single-study runtime seed:

| Test | Setup | Assertion |
| --- | --- | --- |
| `noPagination_returnsAllStudies` | `pageSize=null` | Returns all 25 |
| `pageNumberZero_treatedAsFirstPage` | `pageSize=10, pageNumber=0` | Returns study-0..9 |
| `pageSizeOnly_defaultsToFirstPage` | `pageSize=10, pageNumber=null` | Returns study-0..9 |
| `pageNumber1_returnsFirstPage` | `pageSize=10, pageNumber=1` | Returns study-0..9 |
| `pageNumber2_returnsSecondPage` (primary regression) | `pageSize=10, pageNumber=2` | Returns study-10..19 |
| `lastPartialPage_returnsRemainingStudies` | `pageSize=10, pageNumber=3` | Returns study-20..24 (5 items, no IOOBE) |
| `pageNumberBeyondEnd_returnsEmptyList` | `pageSize=10, pageNumber=4` | Returns empty |

</details>

### Follow-Ups

| Type | Item | Blocking |
| --- | --- | --- |
| Merge requirement | GitHub mergeability state is `BLOCKED` — must be resolved before merge | Yes (merge) |
| Non-blocking | Repository-level pagination gap: `ClickhouseCancerStudyRepository` also ignores `pageNumber` (controller `subList` works around this). Explicitly deferred per PR description (`@PostFilter` authorization constraint). Track as a separate ticket. | No |
| Informational | `X-Total-Count` header for META projection reports paginated count, not true total — pre-existing design limitation, unchanged by this PR | No |
| Informational | Runtime seed has 1 study (`study_es_0`); multi-study offset correctness is unit-test-proven, not runtime-proven | No |

### Recommendation

**Approve with notes.** The pagination bug is definitively fixed: three independent runtime scenarios prove the bug on base and the fix on head, four backward-compatibility paths show zero contract drift, and seven unit tests cover multi-study offset arithmetic unreachable by the runtime seed. The change is a single-method, read-only controller adjustment with no writes, no downstream effects, and minimal blast radius. Resolve the GitHub mergeability block before merging.

**Comment by Pradyuman-aviator** [comment]
@dippindots Please have a Look!

**Comment by Pradyuman-aviator** [comment]
@inodb  Please have a look if you have time....Thanks!!!

**Comment by dippindots** [comment]
@Pradyuman-aviator Will review this soon

File: src/main/java/org/cbioportal/application/rest/vcolumnstore/ColumnStoreStudyController.java [hunk 1]
```diff
@@ -118,10 +122,18 @@ public ResponseEntity<List<CancerStudyMetadataDTO>> getAllStudies(
...
+    // because the authorized study set is only known after @PostFilter runs.
+    //
+    // pageNumber is 1-based. Values <= 0 and null are treated as page 1 (offset 0) for backward
+    // compatibility. Long arithmetic prevents overflow when pageSize * pageNumber is large.
     if (pageSize != null) {
-      studies = studies.stream().limit(pageSize).toList();
+      long offset =
+          (pageNumber != null && pageNumber > 0) ? (long) pageSize * (pageNumber - 1) : 0L;
+      int fromIndex = (int) Math.min(offset, studies.size());
+      int toIndex = (int) Math.min(offset + pageSize, studies.size());
```
[the remark this answers is not in the document]
Comment (Pradyuman-aviator) [reply src/main/java/org/cbioportal/application/rest/vcolumnstore/ColumnStoreStudyController.java:147]: Fixed..... totalCount is now captured from studies.size() before the subList is applied and the X-Total-Count / total-count headers use that pre pagination value. Added a regression test (getAllStudies_metaProjectionWithPagination_totalCountHeaderReportsFullCount) to cover this case.

**Comment by dippindots** [comment]
@Pradyuman-aviator Thanks for openning a PR, we will take a look at it this week

**Comment by Pradyuman-aviator** [comment]
Hey @dippindots! Just rebased onto master to resolve the merge conflict. The conflict was an add/add on ColumnStoreStudyControllerTest.java PR #12273 (by @oleguerplantalech) had added tests for the readPermission fix and this branch had the pagination tests, both creating the same file.

Resolved by merging both test sets into one file ..now has 10 tests total: 2 for read-permission mapping and 8 for pagination. All passing.

**Review by dippindots — APPROVED** [review]
LGTM, we ususally handle pagination in the mapper layer but since studies endpoint has less data and it can be cached by fetching all studies at once, we think it's ok to have this in the controller layer.
